pkg/utils/dataset.go

```diff
@@ -89,3 +97,82 @@ func IsTargetPathUnderFluidNativeMounts(targetPath string, dataset datav1alpha1.
 
 	return false
 }
+
+// GetUFSToUpdate get a UFSToUpdate of mount points to be added and removed
+func GetUFSToUpdate(dataset *datav1alpha1.Dataset) (ufsToUpdate UFSToUpdate) {
+	// get spec MountPaths and mounted MountPaths from dataset
+	specMountPaths, mountedMountPaths := getMounts(dataset)
+	// get mount points which are needed to be added/removed
+	ufsToUpdate.ToAdd, ufsToUpdate.ToRemove = calculateMountPointsChanges(specMountPaths, mountedMountPaths)
```

Comment:
Provide a proposed implementation
```go
type UFSToUpdate struct {
    ToAdd    []string
    ToRemove []string
}

func (u *UFSToUpdate) CalPathsDelta(ds *datav1alpha1.Dataset){
         for _, mount := range dataset.Spec.Mounts {
		if common.IsFluidNativeScheme(mount.MountPoint) {
			continue
		}
		m := UFSPathBuilder{}.GenAlluxioMountPath(mount, dataset.Spec.Mounts)
		desireMountPaths = append(desireMountPaths, m)
	}
	for _, mount := range dataset.Status.Mounts {
		if common.IsFluidNativeScheme(mount.MountPoint) {
			continue
		}
		m := UFSPathBuilder{}.GenAlluxioMountPath(mount, dataset.Status.Mounts)
		mountedMountPaths = append(currentMountPaths, m)
	}

     u.ToAdd = SubtractString(desireMountPaths, currentMountPaths)
     u.ToRemove = SubtractString(currentMountPaths, desireMountPaths)
}
```

Reply:
Thanks for @allenhaozi 's great suggestions! Can we make the methods to encapsulate into `UFSToUpdate`.

```
type UFSToUpdate struct {
    toAdd    []string
    toRemove []string
    dataset *datav1alpha1.Dataset
}

func NewUFSToUpdate(ds *datav1alpha1.Dataset) *UFSToUpdate{
	return &UFSToUpdate{
		dataset: ds,
	}
}

func (u *UFSToUpdate) AnalyzePathsDelta(){
         for _, mount := range u.dataset.Spec.Mounts {
		if common.IsFluidNativeScheme(mount.MountPoint) {
			continue
		}
		m := UFSPathBuilder{}.GenAlluxioMountPath(mount, u.dataset.Spec.Mounts)
		desireMountPaths = append(desireMountPaths, m)
	}
	for _, mount := range u.dataset.Status.Mounts {
		if common.IsFluidNativeScheme(mount.MountPoint) {
			continue
		}
		m := UFSPathBuilder{}.GenAlluxioMountPath(mount, u.dataset.Status.Mounts)
		mountedMountPaths = append(currentMountPaths, m)
	}

     u.toAdd = SubtractString(desireMountPaths, currentMountPaths)
     u.toRemove = SubtractString(currentMountPaths, desireMountPaths)
}

func (u UFSToUpdate)ShouldUpdate() bool {
	return len(u.toAdd) > 0 && len(u.toRemove) > 0
}

func (u UFSToUpdate)ToAdd() []string {
	return u.toAdd
}

func (u UFSToUpdate)ToRemove() []string {
	return u.toRemove
}


```

Reply:
> Thanks for @allenhaozi 's great suggestions! Can we make the methods to encapsulate into `UFSToUpdate`.
> 
> ```
> type UFSToUpdate struct {
>     toAdd    []string
>     toRemove []string
>     dataset *datav1alpha1.Dataset
> }
> 
> func NewUFSToUpdate(ds *datav1alpha1.Dataset) *UFSToUpdate{
> 	return &UFSToUpdate{
> 		dataset: ds,
> 	}
> }
> 
> func (u *UFSToUpdate) AnalyzePathsDelta(){
>          for _, mount := range u.dataset.Spec.Mounts {
> 		if common.IsFluidNativeScheme(mount.MountPoint) {
> 			continue
> 		}
> 		m := UFSPathBuilder{}.GenAlluxioMountPath(mount, u.dataset.Spec.Mounts)
> 		desireMountPaths = append(desireMountPaths, m)
> 	}
> 	for _, mount := range u.dataset.Status.Mounts {
> 		if common.IsFluidNativeScheme(mount.MountPoint) {
> 			continue
> 		}
> 		m := UFSPathBuilder{}.GenAlluxioMountPath(mount, u.dataset.Status.Mounts)
> 		mountedMountPaths = append(currentMountPaths, m)
> 	}
> 
>      u.toAdd = SubtractString(desireMountPaths, currentMountPaths)
>      u.toRemove = SubtractString(currentMountPaths, desireMountPaths)
> }
> 
> func (u UFSToUpdate)ShouldUpdate() bool {
> 	return len(u.toAdd) > 0 && len(u.toRemove) > 0
> }
> 
> func (u UFSToUpdate)ToAdd() []string {
> 	return u.toAdd
> }
> 
> func (u UFSToUpdate)ToRemove() []string {
> 	return u.toRemove
> }
> ```
```
func (u UFSToUpdate)ShouldUpdate() bool {
	return len(u.toAdd) > 0 || len(u.toRemove) > 0
}
```